core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -361,14 +364,19 @@ public ReserveResult tryReserve(PendingAcquire acquire)
                         }
                         return false;
                     })
-                    .max(comparing(node -> nodesRemainingMemory.get(node.getNodeIdentifier())));
+                    .max(comparing(node -> nodesUnreservedMemory.get(node.getNodeIdentifier())));
 
             if (selectedNode.isEmpty()) {
                 return ReserveResult.NONE_MATCHING;
             }
 
             boolean selectedCandidateMatches = false;
-            if (nodesRemainingMemory.get(selectedNode.get().getNodeIdentifier()) >= acquire.getMemoryLease() || isNodeEmpty(selectedNode.get())) {
+
+            String selectNodeId = selectedNode.get().getNodeIdentifier();
+            boolean enoughUnreservedMemory = nodesUnreservedMemory.get(selectNodeId) >= acquire.getMemoryLease();
+            boolean enoughRuntimeMemory = !nodeMemoryPoolInfos.containsKey(selectNodeId)
```

Comment:
why not `nodeMemoryPoolInfos.containsKey(selectNodeId) && nodeMemoryPoolInfos.get(selectNodeId).getFreeBytes() >= acquire.getMemoryLease()`?

Reply:
That is a question if are pessimistic or optimistic in case we do not have info about memory pool usage for given node. Honestly it probably does not matter much. I can change if you prefer :)